### DISCUSSION TITLE
[Question] How to configure different clock frequency between systembus and periphery bus?

### DISCUSSION THREAD

**Discussion by RaulGoal** [chipsalliance/rocket-chip#3215] [Q&A]
Hi,
I have added a simple periphery module on periphery bus and executed c simulation.
I intend to make clock frequency of periphery bus distinguish from system bus.

I try and modify below  frequency value to 5000000 for achieving my goal.
https://github.com/freechipsproject/rocket-chip/blob/a48dd575b23aa4b82368f19beaaf19e2d4bf1e3e/src/main/scala/coreplex/PeripheryBus.scala#L18

 I have checked the simulation wave form diagram after such modification. 
![image](https://user-images.githubusercontent.com/34787655/36770526-a10aca30-1c85-11e8-8ae8-e2dd5b235519.png)
However, periphery bus clock frequency remains same as system bus.

I think that I use wrong approach to achieve my goal.
Can anyone tells me how to set it or the method?

**Comment by terpstra** [comment]
The field you changed is only used to describe the frequency to the SoC. Actual clock generation does not happen in Chisel at all. At SiFive we generally get our clocks from an external oscillator. For simulation, the test-bench provides the clock. If you are using 'emulator/' then the clock comes from csrc/emulator.cc.

**Comment by RaulGoal** [comment]
Thanks for your reply.
I have looked the csrc/emulator.cc file. It probably have to generate the othe clock as `tile->clk`  for periphery bus. However, it might also need to some modifications or addition to Chisel implement code
for identifying the module work at clk frequency.
ig:
`class PeripheryBus(params: PeripheryBusParams)(implicit p: Parameters) extends TLBusWrapper(params, "PeripheryBus") {                                                                                                                                            `
 `clock :=clk  //secify clock frquency = clk`
 `...........}`

I am not familiar with such modification.
How to do that?
Can you give me some suggestions or tips?

**Comment by terpstra** [comment]
Just FYI, no one has done what you are trying to do and it's highly unlikely we would merge changes to this effect back into emulator.cc. That said, your best bet is probably to move the tiles into their own clock domain and plumb signals out of the design for them. Shoving the Periphery is something I know @hcook is working to make easier, but AFAIK this work is not completed.

**Comment by hcook** [comment]
#1190 adds support for specifying a clock crossing between the system bus and periphery bus, but doesn't actually create a different clock wired to the outside. You could do that manually. The hard part is that then all the slave devices attached to the periphery bus had better be put on its same clock... I think we'll have better support for this in a month or two.

**Comment by codelec** [comment]
so am i correct on this that periphery bus means bus supporting only TL-UL type of transactions and not an under-clocked bus for slower peripherals?

**Comment by GitZgl** [comment]
I am aslo confused that how can I generate a clock crossing SOC. In this Soc, system bus running on a fast clock, and periphery bus running on a slow clock.
